src/backend/distributed/utils/statistics_collection.c

```diff
@@ -0,0 +1,213 @@
+/*-------------------------------------------------------------------------
+ *
+ * statistics_collection.c
+ *	  Anonymous reports and statistics collection.
+ *
+ * Copyright (c) 2017, Citus Data, Inc.
+ *
+ *-------------------------------------------------------------------------
+ */
+
+#if HAVE_LIBCURL
+
+#include "postgres.h"
+
+#include <curl/curl.h>
+#include <sys/utsname.h>
+
+#include "access/xact.h"
+#include "citus_version.h"
+#include "distributed/metadata_cache.h"
+#include "distributed/statistics_collection.h"
+#include "distributed/worker_manager.h"
+#include "lib/stringinfo.h"
+#include "utils/json.h"
+
+bool EnableStatisticsCollection = true; /* send basic usage statistics to Citus */
+
+static uint64 NextPow2(uint64 n);
+static uint64 ClusterSize(List *distributedTableList);
+static bool SendHttpPostJsonRequest(const char *url, const char *postFields, long
+									timeout_seconds);
+
+/* WarnIfSyncDNS warns if libcurl is compiled with synchronous DNS. */
+void
+WarnIfSyncDNS(void)
+{
+	curl_version_info_data *version_info = curl_version_info(CURLVERSION_NOW);
+	if (!(version_info->features & CURL_VERSION_ASYNCHDNS))
+	{
+		ereport(WARNING, (errmsg("your current libcurl version doesn't support "
+								 "asynchronous DNS, which might cause unexpected "
+								 "delays in database's operation"),
+						  errhint("Install a libcurl version with asynchronous DNS "
+								  "support.")));
+	}
+}
+
+
+/*
+ * CollectBasicUsageStatistics sends basic usage statistics to Citus servers.
+ * This includes Citus version, table count rounded to next power of 2, cluster
+ * size rounded to next power of 2, worker node count, and uname data. Returns
+ * true if we actually have sent statistics to the server.
+ */
+bool
+CollectBasicUsageStatistics(void)
+{
+	List *distributedTables = NIL;
+	uint64 roundedDistTableCount = 0;
+	uint64 roundedClusterSize = 0;
+	uint32 workerNodeCount = 0;
+	StringInfo fields = makeStringInfo();
+	struct utsname unameData;
+	memset(&unameData, 0, sizeof(unameData));
+
+	StartTransactionCommand();
+	distributedTables = DistributedTableList();
+	roundedDistTableCount = NextPow2(list_length(distributedTables));
+	roundedClusterSize = NextPow2(ClusterSize(distributedTables));
+	workerNodeCount = ActivePrimaryNodeCount();
+	CommitTransactionCommand();
+
+	uname(&unameData);
+
+	appendStringInfoString(fields, "{\"citus_version\": ");
```

Comment:
I still find this pretty hard to read and worry it will be tough to extend. I was envisioning you'd use `composite_to_json` with a tuple we construct in another method (would be clear pairing of column names and field values, and supports nested composite types), or `json_build_object` (could declare like:

```c
{ key, value,
  key, value,
  key, value }
```

and would be clearer what was being generated)

Reply:
After looking over all the options, it's probably best to use `composite_to_json` if we want to have nested fields, but for now let's go with `json_build_object` with an array of our field names and values.

Reply:
Ok. Keeping it this way for now. I'll look into making this more readable, maybe using ```json_build_object```.